text/0000-orderer-snapshot.md

```diff
@@ -0,0 +1,309 @@
+
+    POST /participation/v1/joinbysnapshot
+
+ or the following CLI command:
+
+    $ osnadmin channel joinbysnapshot -s <snapshot_dir>
+
+There is no need to specify {channelID} when using the above API, as it will be extracted from the snapshot content.
+
+To check the status of the "join by snapshot" operation use the following REST API:
+
+    GET /participation/v1/channels/{channelID}
+
+ or the following CLI command:
+
+    $ osnadmin channel list -c <channelID>
+
+## Guide: Prune block ledger of an existing channel
+[guide-prune-block-ledger-of-an-existing-channel]: #guide-prune-block-ledger-of-an-existing-channel
+
+To prune an existing channel ledger do the following:
+ - create a new snapshot at the desired block number,
+ - remove the OSN from the channel,
+ - join the channel from the snapshot.
+
+This approach has some drawbacks, see [Drawbacks section](#drawbacks).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Changes to the existing API
+
+This RFC adds a new endpoint to the existing **channel participation API** to allow joining a channel by snapshot. Additionally this RFC defines new **ledger management API** and introduces a new endpoint of this API to allow snapshot creation. In future we plan to add at least one more endpoint to this API which will enable ledger pruning "on the fly" without interruption in OSN service to resolve the known [drawbacks](#drawbacks).
+
+Technically we achieve the above by following these steps:
+1. Introduce new REST API endpoint **/participation/v1/joinbysnapshot** while keeping all existing functionality of the channel participation API, that is keeping "join by join-block", "remove channel", and "list channel(s)" API endpoints.
+2. Add new REST API handler for the ledger management API with prefix **/ledger/v1/**. Define a single endpoint of this API **/ledger/v1/{channelID}/snapshot** to allow snapshot creation.
+3. Add "Ledger Management API Configuration" section in the OSN YAML config file.
+4. Add support of two new snapshot operations to the CLI tool *osnadmin*, specifically the commands *osnadmin channel joinbysnapshot* and *osnadmin ledger snapshot*.
+
+## Snapshot content
+
+We propose the OSN snapshot to follow the same logical structure as peer snapshot, that is to comprise a directory with a number of files in it, one of the files being a metadata JSON file. But here the similarity ends, as the content of the files will be drastically different.
+Here is the content of the OSN snapshot:
+1. Snapshot version - to allow for the future extensions. Current version is 1.
+2. Channel ID.
+3. Last config block for the channel.
+4. lastBlockNum - the number of the last pruned block. This can be either behind the last config block, equal to the last config block number, or ahead of it.
```

Comment:
At first I assumed that the lastBlockNum must be higher than the last config block number, so that the orderer could process blocks using the latest config right after bootstrapping from a snapshot. But perhaps this is not needed, since the orderer could simply replicate blocks instead of process blocks, until it reached the specified config block in the future. Can anybody confirm?

Reply:
As far as I understand it, this is exactly what the current channel participation code is doing ("channel join" API). It starts the orderer in a "follower" mode where it just pulls the blocks without processing them until it finds a config block confirming its membership in the channel.
I propose to use exactly the same logic (and the same code) in case of snapshot bootstrapping and start the orderer in a "follower" mode in the following cases:
- lastBlockNum is behind the lastConfigBlock, so the membership status of the orderer at the time of  this last pruned block is unclear (unless we can deduce the membership status from the metadata - can we?)
- lastBlockNum is equal or ahead of lastConfigBlock and the orderer is not a member of the channel (not in the consenter set).

Reply:
I've included the explicit description here:
f177408b823ef964cac27fd876db33cd8385ea9d

Reply:
>lastBlockNum is behind the lastConfigBlock, so the membership status of the orderer at the time of this last pruned block is unclear (unless we can deduce the membership status from the metadata - can we?)

As far as I understand, in the metadata it would only be raft-node-IDs not the actual identities. However, in the "follower" mode to catch up in this scenario, it would need the channel config for verifying the signature. So, most recent channel config below the snapshot block in this case would also be needed. @yacovm and @jyellick certainly are more knowledgeable and would be good to hear from them.

Reply:
In order to pull a block, and properly validate it, you need to have the config block which preceded it.  That config block contains the MSP data (so that you can verify identities), and the BlockValidation policy which allows you take verified sets of identities from the block signatures (in the Raft case, just one), and verify that those identities were authorized to produce the block.

There are some definitely advantages to doing hash chain validation, instead of signature validation, and I believe the code does perform some optimization around this, but fundamentally, unless we modify the block replication to retrieve the entire chain backwards (which, could be done), if the lastBlockNum is sufficiently old, you will need the config before it to validate those blocks.  I would suggest that lastBlockNum must _always_ be greater than or (better yet) equal to the last config block.

Reply:
Done, here:
0a36f369aa3d54a84c99fe2cd7f54d4b0b0cfcb1

A comment:
Config blocks happen not that often, so if we limited lastBlockNum to be always a config block, we would loose the ability to prune the ledger at a desired point. Hence lastBlockNum is allowed to be greater than the last config block sequence number.